File: backend/src/app/schemas/track.py
from pydantic import BaseModel, Field
from uuid import UUID
from datetime import datetime
from typing import Optional, List, Literal, Union
from app.utils.enums import DifficultyLevel
from app.schemas.resource import ResourceCreate


class ExistingResourceRef(BaseModel):
    kind: Literal["existing"] = "existing"
    resource_id: str
    position: int


class NewResourceItem(BaseModel):
    kind: Literal["new"] = "new"
    resource: ResourceCreate
    position: int

TrackResourceItem = Union[ExistingResourceRef, NewResourceItem]


class TrackCreate(BaseModel):
    title: str
    short_description: str
    level:DifficultyLevel
    skills: List[str] = Field(default_factory=list) 
    estimated_time: Optional[str] = None
    image_url: Optional[str] = None
    resources: List[TrackResourceItem] = Field(default_factory=list, min_length=1)
    created_by_user_id: Optional[UUID] = None  # Set by backend from authenticated user


class TrackUpdate(BaseModel):
    title: Optional[str] = None
    short_description: Optional[str] = None
    level: DifficultyLevel
    skills: List[str] = Field(default_factory=list)
    estimated_time: Optional[str] = None
    image_url: Optional[str] = None
    resources: List[TrackResourceItem] = Field(default_factory=list)


class ResourceSummary(BaseModel):
    id: UUID
    title: str
    short_description: str
    type: str
    position: int
    level: DifficultyLevel
    skills: List[str] = Field(default_factory=list)
    estimated_time: Optional[str] = None
    image_url: Optional[str] = None

    class Config:
        from_attributes = True


class TrackNameItem(BaseModel):
    """Lightweight track item for dropdowns/autocomplete."""
    id: UUID
    title: str

    class Config:
        from_attributes = True


class TrackRead(BaseModel):
    """Track metadata without resources."""
    id: UUID
    title: str
    short_description: str
    level: DifficultyLevel
    skills: List[str] = Field(default_factory=list)
    estimated_time: Optional[str] = None
    image_url: Optional[str] = None
    created_by_user_id: str
    created_at: datetime

    class Config:
        from_attributes = True


class TrackReadWithResources(TrackRead):
    """Track with full details including resources."""
    resources: List[ResourceSummary] = Field(default_factory=list, min_length=1)

File: frontend/src/types/track.ts
import { TrackLevel } from '../constants/backend';

/** Common track fields shared across all track types */
export interface TrackBase {
  title: string;
  short_description: string;
  level: TrackLevel;
  estimated_time?: string;
  image_url?: string;
  skills: string[];
}

/** Reference to an existing resource */
export interface ExistingResourceRef {
  kind: 'existing';
  resource_id: string;
  position: number;
}

/** New resource data to be created and linked to track */
export interface NewResourceItem {
  kind: 'new';
  resource: {
    title: string;
    short_description: string;
    url: string;
    platform: string;
    resource_type: string;
    level: TrackLevel;
    estimated_time?: string;
    image_url?: string;
    skills?: string[];
    default_funding_type: string;
  };
  position: number;
}

/** Track resource item - either existing resource or new resource */
export type TrackResourceItem = ExistingResourceRef | NewResourceItem;

/** Payload for creating a new track */
export interface TrackCreate extends TrackBase {
  resources: TrackResourceItem[]; // resources to add to track
}

/** Track returned from the API */
export interface TrackRead extends TrackBase {
  id: string;
  skills: string[]; // Override: always present (non-optional)
  image_url: string; // Override: always present (non-optional)
  created_at?: string;
  updated_at?: string;
}

/** Track returned from the API with full resource details */
export interface TrackReadWithResources extends TrackRead {
  resources: ResourceSummary[];
}

/** Payload for updating a track (all fields optional) */
export type TrackUpdate = Partial<TrackBase>;

/** Track name item for dropdowns/autocomplete */
export interface TrackNameItem {
  id: string;
  title: string;
}

/** Resource summary within a track */
export interface ResourceSummary {
  id: string;
  title: string;
  short_description: string;
  url: string;
  platform: string;
  type: string;
  level: TrackLevel;
  skills: string[];
  estimated_time?: string;
  image_url?: string;
  position: number;
}

/** Track resource position item (for internal use) */
export interface TrackResourcePosition {
  id: string;
  position: number;
}

/** UI display representation of a track (transformed from TrackRead) */
export interface TrackDisplay {
  id: string;
  title: string;
  description: string;
  image: string;
  level: TrackLevel;
  estimated_time?: string;
  skills: string[];
  resourceCount?: number;
}

/** Track list response with pagination info */
export interface TrackListResponse {
  items: TrackRead[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

/** Track list filter parameters */
export interface TrackListParams {
  search?: string;
  skill?: string[];
  level?: string[];
  page?: number;
  page_size?: number;
}
